core/src/main/java/com/nextdoor/bender/ipc/IpcSenderService.java

```diff
@@ -152,12 +157,19 @@ public void uncaughtException(Thread th, Throwable ex) {
   synchronized public void flush() throws InterruptedException, TransportException {
     synchronized (buffers) {
       /*
-       * Send what remains in the buffers
+       * Send what remains in the buffers.
+       * If there are errors, we will immediately send an error after this sync block.
        */
       this.buffers.forEach((partition, buffer) -> send(buffer, partition));
       this.buffers.clear();
     }
 
+    if (this.hasUnrecoverableException.getAndSet(false)) {
```

Comment:
I'm unclear why you added this hear. We do the same check below as the last step because we need to wait for threads to stop.

Reply:
Right yeah we should add this after the while loops confirms threads are stopped. But this check should be done before the flush since we don't want to have partial success of aborting one addEvent and completing the rest. The handler to my understanding should be all or nothing.

Reply:
Right be we still need to wait and get runtime stats even on a failure.

Reply:
@stlava right that's fine. We should move the flush as the last statement and do this check for hasUnrecoverable exception right before so we ensure flush is being called with expectation that everything succeeded. This should be fine since the factory.close() method doesn't interact with any of threads/stats.

Let me know if this proposed flow is fine:
1. Send last of events in buffer
2. Wait for threads to complete
3. Collect runtime stats of threads
4. Check for hasUnrecoverableException
5. close factory